src/poslenstrings.jl

```diff
@@ -0,0 +1,526 @@
+    u |= UInt32(b) << 16
+    # second continuation byte
+    ((i += 1) > n) | (u < 0xe0000000) && @goto ret
+    @inbounds b = codeunit(s, i)
+    b & 0xc0 == 0x80 || @goto ret
+    u |= UInt32(b) << 8
+    # third continuation byte
+    ((i += 1) > n) | (u < 0xf0000000) && @goto ret
+    @inbounds b = codeunit(s, i)
+    b & 0xc0 == 0x80 || @goto ret
+    u |= UInt32(b); i += 1
+@label ret
+    return reinterpret(Char, u), i
+end
+
+Base.@propagate_inbounds function Base.getindex(s::PosLenString, i::Int)
+    b = codeunit(s, i)
+    u = UInt32(b) << 24
+    Base.between(b, 0x80, 0xf7) || return reinterpret(Char, u)
+    return getindex_continued(s, i, u)
+end
+
+function getindex_continued(s::PosLenString, i::Int, u::UInt32)
+    if u < 0xc0000000
+        # called from `getindex` which checks bounds
+        @inbounds isvalid(s, i) && @goto ret
+        Base.string_index_err(s, i)
+    end
+    n = ncodeunits(s)
+
+    (i += 1) > n && @goto ret
+    @inbounds b = codeunit(s, i) # cont byte 1
+    b & 0xc0 == 0x80 || @goto ret
+    u |= UInt32(b) << 16
+
+    ((i += 1) > n) | (u < 0xe0000000) && @goto ret
+    @inbounds b = codeunit(s, i) # cont byte 2
+    b & 0xc0 == 0x80 || @goto ret
+    u |= UInt32(b) << 8
+
+    ((i += 1) > n) | (u < 0xf0000000) && @goto ret
+    @inbounds b = codeunit(s, i) # cont byte 3
+    b & 0xc0 == 0x80 || @goto ret
+    u |= UInt32(b)
+@label ret
+    return reinterpret(Char, u)
+end
+
+Base.length(s::PosLenString) = length_continued(s, 1, ncodeunits(s), ncodeunits(s))
+
+Base.@propagate_inbounds function Base.length(s::PosLenString, i::Int, j::Int)
+    @boundscheck begin
+        0 < i ≤ ncodeunits(s)+1 || throw(BoundsError(s, i))
+        0 ≤ j < ncodeunits(s)+1 || throw(BoundsError(s, j))
+    end
+    j < i && return 0
+    @inbounds i, k = thisind(s, i), i
+    c = j - i + (i == k)
+    length_continued(s, i, j, c)
+end
+
+@inline function length_continued(s::PosLenString, i::Int, n::Int, c::Int)
+    i < n || return c
+    @inbounds b = codeunit(s, i)
+    @inbounds while true
+        while true
+            (i += 1) ≤ n || return c
+            0xc0 ≤ b ≤ 0xf7 && break
+            b = codeunit(s, i)
+        end
+        l = b
+        b = codeunit(s, i) # cont byte 1
+        c -= (x = b & 0xc0 == 0x80)
+        x & (l ≥ 0xe0) || continue
+
+        (i += 1) ≤ n || return c
+        b = codeunit(s, i) # cont byte 2
+        c -= (x = b & 0xc0 == 0x80)
+        x & (l ≥ 0xf0) || continue
+
+        (i += 1) ≤ n || return c
+        b = codeunit(s, i) # cont byte 3
+        c -= (b & 0xc0 == 0x80)
+    end
+end
+
+# custom array type that returns PosLenStrings on index
+# copy:
+  # calcs total len of all strings in poslens
+  # allocates fresh data byte buffer for total len
+  # loops through and copies each strings data into new buffer, recalcs poslen
+"""
+    PosLenStringVector{T}(data, poslens, e=UInt8('\\'))
+
+Custom array 
+"""
+struct PosLenStringVector{T} <: AbstractVector{T}
```

Comment:
why does this type have a free parameter `T`?

Reply:
This allows callers/users to specify whether the element type is `Union{Missing, PosLenString}` or just plain `PosLenString`. in CSV.jl, we keep track if any `missing` values were detected while parsing, so we can set element type to just `PosLenString` for better type stability when calling `getindex`